### PR TITLE
[Merged by Bors] - feat(RingTheory/Derivation/MapCoeffs): define `mapCoeffs` on a differential ring

### PR DESCRIPTION
**Pull request by Command-Master**
---

- [x] depends on: #14966
- [x] depends on: #14965 
- [x] depends on: #14699 
- [x] depends on: #14702 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Derivation/MapCoeffs.lean [hunk 1]
```diff
@@ -99,3 +101,87 @@ theorem apply_eval_eq (x : A) (p : A[X]) :
...
+variable [Algebra.IsSeparable A R]
+
+lemma algHom_deriv' (f : R →ₐ[A] R') (hf : Function.Injective f) (x : R) :
+    f (x′) = (f x)′ := algHom_deriv f hf x (Algebra.IsSeparable.isSeparable' x)
+
+omit [Nontrivial R] in
+lemma algEquiv_deriv' (f : R ≃ₐ[A] R') (x : R) :
+    f (x′) = (f x)′ :=
+  haveI := f.nontrivial
+  algHom_deriv' f.toAlgHom f.injective x
```
Comment (Command-Master) [inline Mathlib/RingTheory/Derivation/MapCoeffs.lean:185]: I'm uncertain about having these results here, perhaps they belong in `FieldTheory`

File: Mathlib/RingTheory/Derivation/MapCoeffs.lean [hunk 2]
```diff
@@ -99,3 +101,117 @@ theorem apply_eval_eq (x : A) (p : A[X]) :
...
+variable {A : Type*} [CommRing A] [Differential A]
+
+/--
+A specialization of `Derivation.mapCoeffs` for the case of a differential ring.
+-/
+def mapCoeffs : Derivation ℤ A[X] A[X] :=
+  PolynomialModule.equivPolynomialSelf.compDer Differential.deriv.mapCoeffs
+
+@[simp]
+lemma mapCoeffs_apply (p : A[X]) (i) :
```
Comment (jcommelin) [inline Mathlib/RingTheory/Derivation/MapCoeffs.lean:116]:
```suggestion
lemma coeff_mapCoeffs (p : A[X]) (i) :
```

File: Mathlib/RingTheory/Derivation/MapCoeffs.lean (hunk 1 above)
Comment (jcommelin) [reply Mathlib/RingTheory/Derivation/MapCoeffs.lean:215]: I think it's probably fine here.

File: Mathlib/RingTheory/Derivation/MapCoeffs.lean [hunk 3]
```diff
@@ -99,3 +101,117 @@ theorem apply_eval_eq (x : A) (p : A[X]) :
...
+
+@[simp]
+lemma implicitDifferential_C (v : A[X]) (b : A) :
+    letI := implicitDifferential v
+    (C b)′ = C b′ := implicitDeriv_C v b
+
+@[simp]
+lemma implicitDifferential_X (v : A[X]) :
+    letI := implicitDifferential v
+    X′ = v := implicitDeriv_X v
```
Comment (jcommelin) [inline Mathlib/RingTheory/Derivation/MapCoeffs.lean:177]: I suggest leaving this block out of this PR, until you need them for something. The `letI` in the statements suggests to me that these will be somewhat tricky to use, and we might converge onto a more usable design late.

Comment (Command-Master) [reply Mathlib/RingTheory/Derivation/MapCoeffs.lean:177]: Alright, I'll add this in a later PR

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**